common/app/redux/fetch-challenges-epic.js

```diff
@@ -49,39 +62,84 @@ export function fetchChallengeEpic(actions, { getState }, { services }) {
     });
 }
 
-export function fetchChallengesEpic(
+export function fetchChallengesForBlockEpic(
   actions,
   { getState },
   { services }
 ) {
   return actions::ofType(
     types.appMounted,
-    types.updateChallenges
+    types.updateChallenges,
+    types.fetchNewBlock.start
   )
-    .flatMapLatest(() => {
-      const lang = langSelector(getState());
+    .flatMapLatest(({ type, payload }) => {
+      const fetchAnotherBlock = type === types.fetchNewBlock.start;
+      const state = getState();
+      let { block: blockName } = challengeSelector(state);
+      const lang = langSelector(state);
+
+      if (fetchAnotherBlock) {
+        const fullBlocks = fullBlocksSelector(state);
+        if (fullBlocks.includes(payload)) {
+          return Observable.of({ type: 'NULL'});
+        }
+        blockName = payload;
+      }
+
       const options = {
-        params: { lang },
-        service: 'map'
+        params: { lang, blockName },
+        service: 'challenge'
       };
       return services.readService$(options)
         .retry(3)
-        .map(({ entities, ...res }) => ({
-          entities: shapeChallenges(
-            entities,
-            isDev
-          ),
-          ...res
-        }))
-        .map(({ entities, result } = {}) => {
-          return fetchChallengesCompleted(
-            entities,
-            result
-          );
-        })
+        .map(fetchChallengesCompleted)
         .startWith({ type: types.fetchChallenges.start })
         .catch(createErrorObservable);
     });
 }
 
-export default combineEpics(fetchChallengeEpic, fetchChallengesEpic);
+function fetchChallengesForNextBlockEpic(action$, { getState }) {
+  return action$::ofType(challenge.checkForNextBlock)
+    .map(() => {
+      let nextChallenge = {};
+      let isNewBlock = false;
+      let isNewSuperBlock = false;
+      const state = getState();
+      const challenge = currentChallengeSelector(state);
+      const superBlocks = superBlocksSelector(state);
+      const entities = entitiesSelector(state);
+      nextChallenge = getNextChallenge(challenge, entities, { isDev });
+      // block completed.
+      if (!nextChallenge) {
+        isNewBlock = true;
+        nextChallenge = getFirstChallengeOfNextBlock(
+          challenge,
+          entities,
+          { isDev }
+        );
+      }
+      // superBlock completed
+      if (!nextChallenge) {
+        isNewSuperBlock = true;
+        nextChallenge = getFirstChallengeOfNextSuperBlock(
+          challenge,
+          entities,
+          superBlocks,
+          { isDev }
+        );
+      }
+      const isNewBlockRequired = (
+        (isNewBlock || isNewSuperBlock) &&
+        !nextChallenge.description
+      );
+      return isNewBlockRequired ?
+        fetchNewBlock(nextChallenge.block) :
+        { type: 'NULL' };
```

Comment:
Instead send just null and use a filter downstream. Don't dispatch and action when you don't need to.